File: tests/abi/pluginbase_layout.cpp
#include "pluginterfaces/base/ipluginbase.h"

#include <cstddef>
#include <cstdio>

#define PRINT_TYPE(Type) std::printf (#Type " size %zu align %zu\n", sizeof (Steinberg::Type), alignof (Steinberg::Type))
#define PRINT_OFFSET(Type, Field) std::printf (#Type "." #Field " offset %zu\n", offsetof (Steinberg::Type, Field))

int main ()
{
	std::printf ("PFactoryInfo.kURLSize %d\n", Steinberg::PFactoryInfo::kURLSize);
	std::printf ("PFactoryInfo.kEmailSize %d\n", Steinberg::PFactoryInfo::kEmailSize);
	std::printf ("PFactoryInfo.kNameSize %d\n", Steinberg::PFactoryInfo::kNameSize);
	std::printf ("PFactoryInfo.kNoFlags %d\n", Steinberg::PFactoryInfo::kNoFlags);
	std::printf ("PFactoryInfo.kClassesDiscardable %d\n", Steinberg::PFactoryInfo::kClassesDiscardable);
	std::printf ("PFactoryInfo.kLicenseCheck %d\n", Steinberg::PFactoryInfo::kLicenseCheck);
	std::printf ("PFactoryInfo.kComponentNonDiscardable %d\n", Steinberg::PFactoryInfo::kComponentNonDiscardable);
	std::printf ("PFactoryInfo.kUnicode %d\n", Steinberg::PFactoryInfo::kUnicode);
	std::printf ("PClassInfo.kManyInstances %d\n", Steinberg::PClassInfo::kManyInstances);
	std::printf ("PClassInfo.kCategorySize %d\n", Steinberg::PClassInfo::kCategorySize);
	std::printf ("PClassInfo.kNameSize %d\n", Steinberg::PClassInfo::kNameSize);
	std::printf ("PClassInfo2.kVendorSize %d\n", Steinberg::PClassInfo2::kVendorSize);
	std::printf ("PClassInfo2.kVersionSize %d\n", Steinberg::PClassInfo2::kVersionSize);
	std::printf ("PClassInfo2.kSubCategoriesSize %d\n", Steinberg::PClassInfo2::kSubCategoriesSize);
	std::printf ("PClassInfoW.kVendorSize %d\n", Steinberg::PClassInfoW::kVendorSize);
	std::printf ("PClassInfoW.kVersionSize %d\n", Steinberg::PClassInfoW::kVersionSize);
	std::printf ("PClassInfoW.kSubCategoriesSize %d\n", Steinberg::PClassInfoW::kSubCategoriesSize);

	PRINT_TYPE (PFactoryInfo);
	PRINT_OFFSET (PFactoryInfo, vendor);
	PRINT_OFFSET (PFactoryInfo, url);
	PRINT_OFFSET (PFactoryInfo, email);
	PRINT_OFFSET (PFactoryInfo, flags);

	PRINT_TYPE (PClassInfo);
	PRINT_OFFSET (PClassInfo, cid);
	PRINT_OFFSET (PClassInfo, cardinality);
	PRINT_OFFSET (PClassInfo, category);
	PRINT_OFFSET (PClassInfo, name);

	PRINT_TYPE (PClassInfo2);
	PRINT_OFFSET (PClassInfo2, cid);
	PRINT_OFFSET (PClassInfo2, cardinality);
	PRINT_OFFSET (PClassInfo2, category);
	PRINT_OFFSET (PClassInfo2, name);
	PRINT_OFFSET (PClassInfo2, classFlags);
	PRINT_OFFSET (PClassInfo2, subCategories);
	PRINT_OFFSET (PClassInfo2, vendor);
	PRINT_OFFSET (PClassInfo2, version);
	PRINT_OFFSET (PClassInfo2, sdkVersion);

	PRINT_TYPE (PClassInfoW);
	PRINT_OFFSET (PClassInfoW, cid);
	PRINT_OFFSET (PClassInfoW, cardinality);
	PRINT_OFFSET (PClassInfoW, category);
	PRINT_OFFSET (PClassInfoW, name);
	PRINT_OFFSET (PClassInfoW, classFlags);
	PRINT_OFFSET (PClassInfoW, subCategories);
	PRINT_OFFSET (PClassInfoW, vendor);
	PRINT_OFFSET (PClassInfoW, version);
	PRINT_OFFSET (PClassInfoW, sdkVersion);

	return 0;
}

File: tools/pluginbase_layout.zig
const std = @import("std");
const pluginbase = @import("vst3-zig").pluginterfaces.base.ipluginbase;

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();
    try stdout.print("PFactoryInfo.kURLSize {}\n", .{pluginbase.PFactoryInfo.kURLSize});
    try stdout.print("PFactoryInfo.kEmailSize {}\n", .{pluginbase.PFactoryInfo.kEmailSize});
    try stdout.print("PFactoryInfo.kNameSize {}\n", .{pluginbase.PFactoryInfo.kNameSize});
    try stdout.print("PFactoryInfo.kNoFlags {}\n", .{pluginbase.PFactoryInfo.kNoFlags});
    try stdout.print("PFactoryInfo.kClassesDiscardable {}\n", .{pluginbase.PFactoryInfo.kClassesDiscardable});
    try stdout.print("PFactoryInfo.kLicenseCheck {}\n", .{pluginbase.PFactoryInfo.kLicenseCheck});
    try stdout.print("PFactoryInfo.kComponentNonDiscardable {}\n", .{pluginbase.PFactoryInfo.kComponentNonDiscardable});
    try stdout.print("PFactoryInfo.kUnicode {}\n", .{pluginbase.PFactoryInfo.kUnicode});
    try stdout.print("PClassInfo.kManyInstances {}\n", .{pluginbase.PClassInfo.kManyInstances});
    try stdout.print("PClassInfo.kCategorySize {}\n", .{pluginbase.PClassInfo.kCategorySize});
    try stdout.print("PClassInfo.kNameSize {}\n", .{pluginbase.PClassInfo.kNameSize});
    try stdout.print("PClassInfo2.kVendorSize {}\n", .{pluginbase.PClassInfo2.kVendorSize});
    try stdout.print("PClassInfo2.kVersionSize {}\n", .{pluginbase.PClassInfo2.kVersionSize});
    try stdout.print("PClassInfo2.kSubCategoriesSize {}\n", .{pluginbase.PClassInfo2.kSubCategoriesSize});
    try stdout.print("PClassInfoW.kVendorSize {}\n", .{pluginbase.PClassInfoW.kVendorSize});
    try stdout.print("PClassInfoW.kVersionSize {}\n", .{pluginbase.PClassInfoW.kVersionSize});
    try stdout.print("PClassInfoW.kSubCategoriesSize {}\n", .{pluginbase.PClassInfoW.kSubCategoriesSize});

    try printType(stdout, "PFactoryInfo", pluginbase.PFactoryInfo);
    try printOffset(stdout, "PFactoryInfo", "vendor", pluginbase.PFactoryInfo, "vendor");
    try printOffset(stdout, "PFactoryInfo", "url", pluginbase.PFactoryInfo, "url");
    try printOffset(stdout, "PFactoryInfo", "email", pluginbase.PFactoryInfo, "email");
    try printOffset(stdout, "PFactoryInfo", "flags", pluginbase.PFactoryInfo, "flags");

    try printType(stdout, "PClassInfo", pluginbase.PClassInfo);
    try printOffset(stdout, "PClassInfo", "cid", pluginbase.PClassInfo, "cid");
    try printOffset(stdout, "PClassInfo", "cardinality", pluginbase.PClassInfo, "cardinality");
    try printOffset(stdout, "PClassInfo", "category", pluginbase.PClassInfo, "category");
    try printOffset(stdout, "PClassInfo", "name", pluginbase.PClassInfo, "name");

    try printType(stdout, "PClassInfo2", pluginbase.PClassInfo2);
    try printOffset(stdout, "PClassInfo2", "cid", pluginbase.PClassInfo2, "cid");
    try printOffset(stdout, "PClassInfo2", "cardinality", pluginbase.PClassInfo2, "cardinality");
    try printOffset(stdout, "PClassInfo2", "category", pluginbase.PClassInfo2, "category");
    try printOffset(stdout, "PClassInfo2", "name", pluginbase.PClassInfo2, "name");
    try printOffset(stdout, "PClassInfo2", "classFlags", pluginbase.PClassInfo2, "classFlags");
    try printOffset(stdout, "PClassInfo2", "subCategories", pluginbase.PClassInfo2, "subCategories");
    try printOffset(stdout, "PClassInfo2", "vendor", pluginbase.PClassInfo2, "vendor");
    try printOffset(stdout, "PClassInfo2", "version", pluginbase.PClassInfo2, "version");
    try printOffset(stdout, "PClassInfo2", "sdkVersion", pluginbase.PClassInfo2, "sdkVersion");

    try printType(stdout, "PClassInfoW", pluginbase.PClassInfoW);
    try printOffset(stdout, "PClassInfoW", "cid", pluginbase.PClassInfoW, "cid");
    try printOffset(stdout, "PClassInfoW", "cardinality", pluginbase.PClassInfoW, "cardinality");
    try printOffset(stdout, "PClassInfoW", "category", pluginbase.PClassInfoW, "category");
    try printOffset(stdout, "PClassInfoW", "name", pluginbase.PClassInfoW, "name");
    try printOffset(stdout, "PClassInfoW", "classFlags", pluginbase.PClassInfoW, "classFlags");
    try printOffset(stdout, "PClassInfoW", "subCategories", pluginbase.PClassInfoW, "subCategories");
    try printOffset(stdout, "PClassInfoW", "vendor", pluginbase.PClassInfoW, "vendor");
    try printOffset(stdout, "PClassInfoW", "version", pluginbase.PClassInfoW, "version");
    try printOffset(stdout, "PClassInfoW", "sdkVersion", pluginbase.PClassInfoW, "sdkVersion");

    try printType(stdout, "IPluginFactory", pluginbase.IPluginFactory);
    try printType(stdout, "IPluginFactory2", pluginbase.IPluginFactory2);
    try printType(stdout, "IPluginFactory3", pluginbase.IPluginFactory3);
}

fn printType(writer: anytype, comptime name: []const u8, comptime Type: type) !void {
    try writer.print("{s} size {} align {}\n", .{ name, @sizeOf(Type), @alignOf(Type) });
}

fn printOffset(writer: anytype, comptime type_name: []const u8, comptime field_label: []const u8, comptime Type: type, comptime field_name: []const u8) !void {
    try writer.print("{s}.{s} offset {}\n", .{ type_name, field_label, @offsetOf(Type, field_name) });
}
